torch_geometric/nn/dense/hoscpool.py

```diff
@@ -0,0 +1,153 @@
+import torch
+
+EPS = 1e-15
+
+
+def dense_hoscpool(
+    x,
+    adj,
+    s,
+    mu=0.1,
+    alpha=0.5,
+    new_ortho=False,
+    mask=None,
+):
+    r"""The highe-order pooling operator (HoscPool) from the paper
+    `"Higher-order clustering and pooling for Graph Neural Networks"
+    <http://arxiv.org/abs/2209.03473>`_. Based on motif spectral clustering,
+    it captures and combines different levels of higher-order connectivity
+    patterns when coarsening the graph.
+
+    .. math::
+        \mathbf{X}^{\prime} &= {\mathrm{softmax}(\mathbf{S})}^{\top} \cdot
+        \mathbf{X}
+
+        \mathbf{A}^{\prime} &= {\mathrm{softmax}(\mathbf{S})}^{\top} \cdot
+        \mathbf{A} \cdot \mathrm{softmax}(\mathbf{S})
+
+    based on the learned cluster assignment matrix :math:`\mathbf{S} \in \mathbb{R}^{B
+    \times N \times K}`. This function returns the pooled feature matrix, the coarsened
+    symmetrically normalised adjacency matrix, the motif spectral clustering loss :math:`\mathcal{L}_{mc}`
+    and the orthogonality loss :math:`\mathcal{L}_{o}`.
+
+    .. math::
+        \mathcal{L}_{mc} &= - \frac{\alpha_1}{K} \cdot \text{Tr}\bigg(\frac{\mathbf{S}^\top \mathbf{A} \mathbf{S}}
+            {\mathbf{S}^\top\mathbf{D}\mathbf{S}}\bigg) - \frac{\alpha_2}{K} \cdot \text{Tr}\bigg(
+                \frac{\mathbf{S}^\top\mathbf{A}_{M}\mathbf{S}}{\mathbf{S}^\top\mathbf{D}_{M}\mathbf{S}}\bigg).
+
+        \mathcal{L}_o &= \frac{1}{\sqrt{K}-1} \bigg( \sqrt{K} - \frac{1}{\sqrt{N}}\sum_{j=1}^K ||S_{*j}||_F\bigg)
+
+    Args:
+        x (Tensor): Node feature tensor :math:`\mathbf{X} \in \mathbb{R}^{B
+            \times N \times F}` with batch-size :math:`B`, (maximum)
+            number of nodes :math:`N` for each graph, and feature dimension
+            :math:`F`.
+        adj (Tensor): adjacency matrix :math:`\mathbf{A} \in \mathbb{R}^{B \times N \times N}`.
+        s (Tensor): the learnable cluster assignment matrix :math:`\mathbf{S} \in \mathbb{R}^{B
+            \times N \times K}` with number of clusters :math:`K`. The softmax
+            does not have to be applied beforehand, since it is executed
+            within this method.
+        mu (Tensor, optional): scalar that controls the importance given to regularization loss
+        alpha (Tensor, optional): scalar in [0,1] controlling the importance granted
+            to higher-order information (in loss function).
+        new_ortho (BoolTensor, optional): either to use new proposed loss or old one
```

Comment:
This argument is not very clear. What is 'new proposed' and 'old one' can we make it a bit more descriptive?